### ISSUE TITLE
lazyproperty: Documentation misleading

### ISSUE THREAD

**Issue by MSeifert04** [astropy/astropy#4471]
I tried the ``lazyproperty`` from the utils.decorators today and it is works as expected but one thing is very strange. The documentation states that

If a setter for this property is defined, it will still be possible to manually update the value of the property, if that capability is desired.

Well but at least with python 3.5 the setter/deleter are always created. To stay with the example in the docstring of the decorator
```
class LazyTest(object):
    @lazyproperty
    def complicated_property(self):
        print('Computing the value for complicated_property...')
        return 42
lt = LazyTest()
lt.complicated_property
#Computing the value for complicated_property...
#42
lt.complicated_property
#42
# New
lt.complicated_property = 10
lt.complicated_property
#10
del lt.complicated_property
lt.complicated_property
#Computing the value for complicated_property...
#42
```
I didn't expected this behaviour from the docstring and was searching for some time now where my code went wrong. If this is intentional I would suggest altering the docstring or if it wasn't intentional maybe there is a way to alter it. I haven't checked but maybe this may also be something that was different in an older python version. Or am I misunderstanding the documentation?

**Comment by mhvk** [comment]
I agree the documentation should be updated; indeed, as we just specifically used the fact that there are default setters and deleters (in #4422), I think we better don't change the behaviour (I now realise we implemented this after noticing there were default setters and deleters by looking at the code -- I never even read the documentation...).

**Comment by pllim** [comment]
@embray might be able to comment on this?

**Comment by embray** [comment]
I wrote `lazyproperty` years ago, so I don't remember exactly what my thinking was on this.  Part of me feels like it would have been better if it were more isomorphic with the `property` built-in, in that setting/deleting of properties is disallowed by default.

Just to guess at my own thinking--the original use cases that I wrote `lazyproperty` for *did* require being able to set the attribute as well (and to a lesser extent delete it).  And as it stands currently writing a setter or deleter for a `lazyproperty` requires some understanding of how `lazyproperty` works internally, in that it stores the cached attribute value in each object's `__dict__`.

Of course, I don't want to change the existing behavior now.  Instead I would suggest that in addition to better documenting the default setter/deleter, the documentation for `lazyproperty` should also be clear about how lazyproperty stores the cached value in each object's `__dict__`.  Although it's in some sense an implementation detail, it's a detail that's important to know about.
